datastore/gcloud/aio/datastore/utils.py

```diff
@@ -1,18 +1,35 @@
+from datetime import datetime as dt
 from typing import Any
 from typing import Dict
 
 from gcloud.aio.datastore.constants import FORMATTERS
 from gcloud.aio.datastore.constants import TypeName
-from gcloud.aio.datastore.constants import TYPES
 from gcloud.aio.datastore.constants import UNFORMATTERS
 
 
+# TODO: add geoPointValue and arrayValue
+# NOTE: TypeName.ENTITY and TypeName.KEY are added dynamically by the datastore
+TYPES = {
+    bool: TypeName.BOOLEAN,
+    bytes: TypeName.BLOB,
+    dt: TypeName.TIMESTAMP,
+    float: TypeName.DOUBLE,
+    int: TypeName.INTEGER,
+    str: TypeName.STRING,
+    type(None): TypeName.NULL,
+}
+
+
 def infer_type(value: Any) -> TypeName:
     kind = type(value)
 
     try:
         return TYPES[kind]
     except KeyError:
+        for supported_type, name in TYPES.items():
```

Comment:
If this is running concurrently against the global `TYPES` and the dictionary changes size after, for example, `TYPES[value] = TypeName.KEY` on a newly created subclass type, does this loop fail?

Reply:
What do you mean by fail?

That the user would be running something that calls this method while updating `entity_result_kind` in Datastore and the would be a race condition and the possibility that the new entity kind would not appear in this loop because it has been updated too late?

I'm not sure what we could do about that.

If your concern is what happens to `items()` if the list is updated externally, then I'm not sure but I'd be surprised if it raised an exception.

Reply:
It would indeed raise an exception: `RuntimeError: dictionary changed size during iteration` in particular. When writing threadsafe code, we absolutely need to avoid modifying globals like this -- you'll need to have this `TYPES` value passed in or stored somewhere threadsafe, instead.

Reply:
Ahhh... My bad.

Well, I don't actually need to support entityValue so I can just remove that and add it later in a different PR. If I remove those lines and the entity_result_kind in DatastoreMeta, then TYPES will be constant, except that the key_kind setter will add the key type. Would that be acceptable?

I think that there is a lot more complexity in adding entityValue, so I shouldn't add it before I need it.

Reply:
I agree wholeheartedly with that -- YAGNI and all that!

I think overall that sounds reasonable? So long as we're avoiding the global multi-thread mutations, we should be good here.

Reply:
Quick note that I think given [this comment](https://github.com/talkiq/gcloud-aio/pull/72#discussion_r281422024), adding `entityValue` support would just be roughly the one-liner of adding `Entity: TypeName.ENTITY` to the `TYPES` dict and changing the `if .. == 'keyValue'` checks to `if .. in {'keyValue', 'entityValue'}`.

Reply:
That would support Entity but not custom entity types. But you're right, that part is very easy to do so I can just add it. Custom entity types can be supported later.

Reply:
Gotcha, did not think of the child classes case. Good call, we can hold off on that for now!